src/graph/visitor/test/ExtractFilterExprVisitorTest.cpp

```diff
@@ -112,5 +191,278 @@ TEST_F(ExtractFilterExprVisitorTest, TestCanBePushSomeAndOr) {
+    ExpressionUtils::pullOrs(expect);
+    expect = getOrExpr(expect);
+    auto remain = getOrs(A, {1, 5, 3, 7});
+
+    checkEqual(expr, expect, remain, true);
+  }
+}
+
+TEST_F(ExtractFilterExprVisitorTest, TestMultiOrExpr) {
+  {
+    // A1 or (A2 and B3)
+    // expect: can push
+    // expect expr: A1 or A2
+    // remain: A1 or B3
+    auto A1 = getCanPushExpr("A", "1");
+    auto A2 = getCanPushExpr("A", "2");
+    auto B3 = getCannotPushExpr("B", "3");
+
+    auto expr = orExpr(A1, andExpr(A2, B3));
+    auto expect = getOrExpr(orExpr(A1, A2));
+    auto remain = orExpr(A1, B3);
+
+    checkEqual(expr, expect, remain, true);
+  }
+  {
+    // A1 or (A2 and B1) or A3 or A4
+    // expect: can push
+    // expect expr: A1 or A3 or A4 or A2
+    // remain: A1 or A3 or A4 or B1
+    std::vector<Expression*> A, B;
+    getExprVec(A, B, 4, 1);
+
+    auto a2b1 = andExpr(A[2], B[1]);
+
+    auto expr1 = orExpr(A[1], a2b1);
+    auto expr = orExpr(expr1, getOrs(A, {3, 4}));
+
+    auto expect = getOrs(A, {1, 3, 4, 2});
+    expect = getOrExpr(expect);
+    auto remain = orExpr(orExpr(A[1], A[3]), orExpr(A[4], B[1]));
+    ExpressionUtils::pullOrs(remain);
+
+    checkEqual(expr, expect, remain, true);
+  }
+  {
+    // A1 or (A2 and B1 and B2 and A3 and B3) or A4 or (A5 and A6)
+    // expect: can push
+    // expect expr: A1 or A4 or (A5 and A6) or (A2 and A3)
+    // remain: A1 or A4 or or (A5 and A6) or (B1 and B2 and B3)
+    std::vector<Expression*> A, B;
+    getExprVec(A, B, 6, 4);
+
+    auto expr1 = getAnds(A, {2, 8, 9, 3, 10});
+    auto expr2 = andExpr(A[5], A[6]);
+    auto expr = orExpr(orExpr(A[1], expr1), orExpr(A[4], expr2));
+
+    auto expect1 = getOrs(A, {1, 4});
+    auto expect2 = orExpr(expect1, orExpr(expr2, andExpr(A[2], A[3])));
+    ExpressionUtils::pullOrs(expect2);
+    auto expect = getOrExpr(expect2);
+    auto remain = orExpr(orExpr(expect1, expr2), getAnds(A, {8, 9, 10}));
+    ExpressionUtils::pullOrs(remain);
+
+    checkEqual(expr, expect, remain, true);
+  }
+}
+
+TEST_F(ExtractFilterExprVisitorTest, TestMultiCanNotPush) {
+  {
+    // A1 or B1 or A2
+    // expect: can not push
+    // expect expr: A1 or B1 or A2
+    // remain: nullptr
+    std::vector<Expression*> A, B;
+    getExprVec(A, B, 2, 1);
+
+    auto expr = orExpr(A[1], orExpr(B[1], A[2]));
+
+    auto expect = expr->clone();
+    ExpressionUtils::pullOrs(expect);
+    auto remain = nullptr;
+
+    checkEqual(expr, expect, remain, false);
+  }
+  {
+    // B1 and B2 and B3
+    // expect: can not push
+    // expect expr: B1 and B2 and B3
+    // remain: nullptr
+    std::vector<Expression*> A, B;
+    getExprVec(A, B, 0, 3);
+
+    auto expr = andExpr(B[1], andExpr(B[2], B[3]));
+
+    auto expect = expr->clone();
+    ExpressionUtils::pullAnds(expect);
+    auto remain = nullptr;
+
+    checkEqual(expr, expect, remain, false);
+  }
+  {
+    // A1 or (A2 and B1) or (A3 and B2) or A4
```

Comment:
it seems that there're some parts of this expression could be pushed:
```
A1 or (A2 and B1) or (A3 and B2) or A4
=>
A1 or A4 or (A2 and B1) or (A3 and B2)
=>
((A1 or A4 or A2) and (A1 or A4 or B1)) or (A3 and B2)
=>
(((A1 or A4 or A2) and (A1 or A4 or B1)) or A3) and (((A1 or A4 or A2) and (A1 or A4 or B1)) or B2)
=>
((A1 or A4 or A2 or A3) and (A1 or A4 or B1 or A3)) and ((A1 or A4 or A2 or B2) and (A1 or A4 or B1 or B2))
=>
(A1 or A4 or A2 or A3) and (A1 or A4 or B1 or A3) and (A1 or A4 or A2 or B2) and (A1 or A4 or B1 or B2)
=>
can be pushed: (A1 or A4 or A2 or A3) 
remain: (A1 or A4 or B1 or A3) and (A1 or A4 or A2 or B2) and (A1 or A4 or B1 or B2)

```

Reply:
remained expr could be simplified: `(A1 or A4) or ((B1 or A3) and (A2 or B2) and (B1 or B2))
`

Reply:
It is not easy to implement this. In this pr, for simplification, LogicalOr can be split, if and only if just one operand can not be pushed and that operand is LogicalAnd.  Maybe this improvement can be implemented in future version.

Reply:
More test cases for different scenarios are necessary to verify the correctness of this pr. However, I can't think of more.

Reply:
make sense. thanks.